pattern/open-source-trumps-innersource

InnerSrcAdmin opened the pull request:
People find the InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with the open source component.

NEXT STEP: Needs to be revised and reviewed
NOTES: Needs further clarification - 2016-10-27

InnerSrcAdmin committed:
Create file

from wiki https://github.com/paypal/InnerSourceCommons/wiki/Open-Source-trumps-InnerSource

gruetter commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,42 @@
+## Statement of Problem
+* People find the InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with the open source component.
+* ORIGINAL: Developers tend to select open source components rather than internally developed components, which results in poorer quality components or also makes it difficult to sustain internal component development.
+* NEW: Different BLs are selecting different SW components to do the same function, resulting in inconsistencies in the user experience.
```
What does _BL_ stand for? Is it business line? Is this different from a _BU_ (business unit)?

NewMexicoKid replied:
BL = Business Line. Typically there are multiple BLs in a Business Unit; but I can guess this varies depending on company.

spier replied:
The BL acronym is spelled out in the Context section now. Resolving.

gruetter commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,42 @@
+
+## Forces 
+* There is a perception that the open source components are higher quality and more reusable. The channels to get any needed changes are more obvious with open source than with internal components. Can easily grab the open source whereas internally developed component has less visibility and ease of availability. (This could be separate Pattern). All things equal, they gravitate towards open sour ...
+* Open source should be voluntary; mandating people to use internally developed software goes against open source principles.
```
I personally don't think this is a strong force. IMHO, voluntariness pertains more to the development of and making contributions to OSS, not necessarily to using it.

lenucksi replied:
Could any of you two (@gruetter @NewMexicoKid) suggest any possible commit suggestion to modify this or resolve this discussion?

gruetter commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,42 @@
+* ORIGINAL: Internal component is well written and there are no quality problems with the software. Whatever component is selected does not need to be forked. The internal component was created and more complete before the open source was available. People are aware there is an internally developed software component.
+* ADDED: People look for software through external search engines.
+
+## Forces 
```
One more possible force in favour of using InnerSource components: licensing. Depending on the regulations within a company, it might be easier to use a _corporate licensed_, InnerSource component than an open source component which is licensed in a way that requires the company to disclose IP.

gruetter replied:
Taken care of in line 22, already.

gruetter commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,42 @@
+## Statement of Problem
+* People find the InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with the open source component.
+* ORIGINAL: Developers tend to select open source components rather than internally developed components, which results in poorer quality components or also makes it difficult to sustain internal component development.
+* NEW: Different BLs are selecting different SW components to do the same function, resulting in inconsistencies in the user experience.
```
Might this be a problem for a dedicated pattern/donut? I feel this is not so much related to OSS trumping InnerSource.

spier replied:
You meant that this one could go to it's own pattern instead, right?
> Different Business Lines (BLs) are selecting different SW components to do the same function, resulting in inconsistencies in the user experience.

Will leave this thread open but I think it should not prevent us from merging this PR.

gruetter commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,42 @@
+
+## Resolution
+Corporate requirement that internally developed components should be evaluated before we go outside to search for open source component. If the open source component is now more mature, replace the internally developed with the open source. Create some extrinsic rewards to encourage them (initially). Make it easier to find the internal component (discoverability and visibility). Make the process  ...
+Provide an internal instance of GitHub Enterprise or an well publicized external GitHub organization repo to allow employees to easily find internally developed solutions. Assign maintainers to make sure proper open source processes are being followed within the leading repos. Provide “value add” services linked to the formal repo solution such as automated CI/CD services, IaaS/PaaS, NPM organiza ...
```
I personally feel that assigning maintainers is the key solution here. I think it deserves a bullet point of its own.

spier replied:
While I don't fully understand what "leading repos" means in this context, @gruetter's feedback has already been incorporated as this sentence is its own bullet now:
> * Assign maintainers to make sure proper open source processes are being followed within the leading repos.

NewMexicoKid committed:
Update open-source-trumps-innersource.md

Cleaned up the pattern. Ready for review #1. 2017-05-10

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,35 @@
+## Title
+Open Source trumps InnerSource
+
```
Interested in a **Patlet** in this pattern?

NewMexicoKid replied:
Yes, it would be good to add:
## Patlet
People find the InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with the open source component.

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,35 @@
+Open Source trumps InnerSource
+
+## Statement of Problem
+* People find the InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with the open source component.
```
It's not clear to me that this behavior is a problem.

NewMexicoKid replied:
I think the core problem is that people's mindset is that open source is always superior to InnerSource, which often is not the case. The result is that this bias leads people to select and use inferior components, perhaps having to put effort to adapting them that would otherwise go towards improving and reusing the InnerSource component.

rrrutledge replied:
I see.  I think that point would be useful to roll into the statement - the inner source component meets their needs better than an open source component.  They should choose the component that best meets their needs irrespective of it being open source vs. inner source?

gruetter replied:
As a tangent: we've seen projects chose expensive COTS over superior InnerSource software, too.

spier replied:
This feedback has been worked into the "Problem" section already. Resolving.

@gruetter if you would like to see your statement about COTS integrated into the pattern, would you mind sending a suggestion for this?

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,35 @@
+
+## Context
+* People look for software through external search engines.
+* Different Business Lines (BLs) are selecting different SW components to do the same function, resulting in inconsistencies in the user experience.
```
This comment doesn't have to be addressed right now, but in the log run it would be nice if our patterns had consistent terminology (glossary) for terms like these (`Business Lines`).

gruetter replied:
An additional effect possibly worth mentioning is that this will also result in more effort for procurement and supplier management.

lenucksi replied:
I've referenced @rrrutledge point into #170.

I've seen @gruetter and @NewMexicoKid agreeing on the possible need to mention the procurement and supplier management point. Can any of you two create a commit suggestion or a commit around this?

spier replied:
Keeping this comment open, to make it easier to find this comment after the PR is merged.

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,35 @@
+## Forces 
+* There is a perception that the open source components are higher quality and more reusable. The channels to get any needed changes are more obvious with open source than with internal components. Can easily grab the open source whereas internally developed component has less visibility and ease of availability. (This could be separate Pattern). All things equal, they gravitate towards open sour ...
+* Open source should be voluntary; mandating people to use internally developed software goes against open source principles.
+* Internally developed components should have advantages over competing open source. If you can clearly demonstrate this, it can be persuasive.
```
This line seems like part-`Context` (`Internally developed components should have advantages over competing open source`) and part-`Solution` (`If you can clearly demonstrate this, it can be persuasive.`) rather than `Forces`.

NewMexicoKid replied:
@rrrutledge how about: Internally developed components can be made more robust and better suited towards corporate needs. Reasons for preferring inner source over open source alternatives might include issues with the open source license, use of proprietary and/or patented components, and functional/competitive advantages. The more that such internal components are used within a company, the better the ROI for the inner source development efforts. On the other hand, open source advances can sometimes leap frog internal development and without periodic assessment, it can be difficult to know when to stop investment in inner source development and switch to the adoption of open source components.

spier replied:
I cannot tell if this thread has been worked into the latest version of the pattern already. Will leave the tread open.

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,35 @@
+* There is a perception that the open source components are higher quality and more reusable. The channels to get any needed changes are more obvious with open source than with internal components. Can easily grab the open source whereas internally developed component has less visibility and ease of availability. (This could be separate Pattern). All things equal, they gravitate towards open sour ...
+* Open source should be voluntary; mandating people to use internally developed software goes against open source principles.
+* Internally developed components should have advantages over competing open source. If you can clearly demonstrate this, it can be persuasive.
+* Part of the reason this is a problem: if different teams are picking different open source components, the situation makes interoperability and integration more difficult; and this could potentially impact the user experience (inconsistency in behaviors).
```
If this is a problem then put it in the `Problem` section.

gruetter replied:
I think this is not the problem this pattern addresses. I will add an additional force for this.

spier replied:
I see this sentence in Context which sounds similar:
> * Different Business Lines (BLs) are selecting different SW components to do the same function, resulting in inconsistencies in the user experience.

Therefore I am assuming that this thread has been resolved.

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,35 @@
+* Open source should be voluntary; mandating people to use internally developed software goes against open source principles.
+* Internally developed components should have advantages over competing open source. If you can clearly demonstrate this, it can be persuasive.
+* Part of the reason this is a problem: if different teams are picking different open source components, the situation makes interoperability and integration more difficult; and this could potentially impact the user experience (inconsistency in behaviors).
+* Depending on the regulations within a company, it might be easier to use a corporate licensed, InnerSource component than an open source component which is licensed in a way that requires the company to disclose IP.
```
Use more clear, definite language (rather than conditional).

gruetter replied:
Proposal: _Using corporate licensed InnerSource components instead of using open source components with strong copyleft license avoids having to disclose IP_.

spier replied:
The last statement has been worked into the pattern already. Resolving thread.

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,35 @@
+* Internally developed components should have advantages over competing open source. If you can clearly demonstrate this, it can be persuasive.
+* Part of the reason this is a problem: if different teams are picking different open source components, the situation makes interoperability and integration more difficult; and this could potentially impact the user experience (inconsistency in behaviors).
+* Depending on the regulations within a company, it might be easier to use a corporate licensed, InnerSource component than an open source component which is licensed in a way that requires the company to disclose IP.
+* There may be many silos in the company; it would then be difficult to reach the developer base across those silos.
```
It's unclear to me what these two last bullet points are trying to get at or how they are uniquely appropriate to describing this pattern?

gruetter replied:
I have the same problem.

NewMexicoKid replied:
I think the point of this force is that it can be difficult to fight perception among dispersed and silo'd engineering staff and to communicate (1) that the inner source project exits, (2) that it has advantages over competing open source, (3) that there are legitimate reasons for preferring the inner source over competing open source software.

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,35 @@
+* It can be hard to find stuff in github (especially if names are cryptic and keywords aren't used). See [Poor Naming Conventions](https://github.com/paypal/InnerSourcePatterns/pull/59).
+
+## Sketch
+See figure 1 in https://drive.google.com/open?id=0B7_9iQb93uBQNlJ0YU5wWmpWYUU 
```
If I understand the diagram (and this article) right, then it looks like the diagram suggests that it is preferable to use an inner source project over an open source project in areas of software that are differentiating to the company's core business?

NewMexicoKid replied:
Yes

lenucksi replied:
What do you think about the following:
- Add this .jpg to the repo maybe in /assets/ or something similar such that it can be rendered with the pattern? Also, who created the image and are there possibly any copyrights attached to it that we need to take into account?
- Should there be some phrasing adding the insight from the image into the text?

spier replied:
The "owner" of the picture in gDrive is @NewMexicoKid and it was created 11.10.2016, 18:33.
Not sure if that helps us to figure out where it originally came frame?

I also did a google image search for the picture and parts of the text, but could find anything.

spier replied:
I found the original source for this image. Yay :)
https://www.semanticscholar.org/paper/Commodification-of-Industrial-Software%3A-A-Case-for-Linden-Lundell/54d6cb77a86e292ff1845eb910c1a1f258e6cee3

rrrutledge committed:
Update open-source-trumps-innersource.md

Putting each sentence on its own line.  Doesn't change the look of the rendered markdown, but makes the document easier to review and comment.

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,40 @@
+See figure 1 in https://drive.google.com/open?id=0B7_9iQb93uBQNlJ0YU5wWmpWYUU 
+
+## Resolution
+* Corporate requirement that internally developed components should be evaluated before we go outside to search for open source component.
```
I doubt the ability for such a requirement to have real, lasting, and actual affect on employee behavior.

gruetter replied:
I have the same reservation.

NewMexicoKid replied:
It may depend on the corporate culture or how this requirement is put in place (there might be a process for finding and evaluating software components or for using open source software). I think it is doable to require this step (we use it in our company).

gruetter replied:
If this is the core solution of this pattern, then I propose we add language to the context which makes it clear that this solution could be successfully applied.

spier replied:
Context contains:
> * The company culture is based on strong governance.

So I think this thread is addressed. Resolving.

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,40 @@
+## Resolution
+* Corporate requirement that internally developed components should be evaluated before we go outside to search for open source component.
+If the open source component is now more mature, replace the internally developed with the open source. Create some extrinsic rewards to encourage them (initially).
+Make it easier to find the internal component (discoverability and visibility).
```
Quite a bit of overlap with `Create a tool with a central view of internally available software`.  Maybe just delete this sentence?

gruetter replied:
I'll add a link to the _InnerSource Portal_ pattern, instead

spier replied:
Thread is worked into the pattern already. Resolving.

rrrutledge commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,40 @@
+## Sketch
+See figure 1 in https://drive.google.com/open?id=0B7_9iQb93uBQNlJ0YU5wWmpWYUU 
+
+## Resolution
```
A lot of these suggested resolutions are (gut-feel) very large undertakings and outside the realm of reasonableness to just go and do.  I agree that they would help the situation your describing, but feel that they are large to the point of their not having practical utility in including them here as independent items.  Perhaps some of them need to be their own pattern and linked to from here?  Or perhaps the need to include such large items as a solution indicates that the problem space that this pattern is trying to address is too broad?

gruetter replied:
@rrrutledge : Are you ok with merging with the current state of the resolution section (there has been a slight rework), or do you think further rework is needed before merging?

spier replied:
While I agree that the **Solutions** section is asking for a lot right now, I suggest that we just leave this as is and get the pattern published as Initial. That will make it easier for others to find, read, and improve this pattern in the future if anybody has actually found a working solution for this problem. (right now we don't have a Known Instance yet, so this isn't a proven solution yet)

rrrutledge reviewed (dismissed):
These thoughts are great, Nick.  I like the broad themes and big-picture items that you've outlined here and am excited to discuss more of the details.

gruetter commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,40 @@
+* Different Business Lines (BLs) are selecting different SW components to do the same function, resulting in inconsistencies in the user experience.
+
+## Forces 
+* There is a perception that the open source components are higher quality and more reusable. The channels to get any needed changes are more obvious with open source than with internal components. Can easily grab the open source whereas internally developed component has less visibility and ease of availability. (This could be separate Pattern). All things equal, they gravitate towards open sour ...
```
I'm not sold on the _channels-are-more-obvious_ argument. In fact, if both the open source and InnerSource project do an equally good job of explaining how to collaborate with them, the InnerSource project might have an advantage (e. g. being able to see the shared calendar of the maintainers).

gruetter replied:
Also, would it make sense to elaborate on where the perception of higher quality comes from? In german, we have a saying that goes something like this: "The prophet doesn't count in is own country".

NewMexicoKid replied:
The perception usually comes from a lack of information, a bias in believing that open source = many more developers and users = higher quality (this is sometimes the case but not always).

spier replied:
I worked in @NewMexicoKid comment as such:
> There is a perception that the open source components are of higher quality and more reusable than the InnerSource components. This is caused by the common belief that open source always means many more developers and users and therefore higher quality. This is sometimes the case in open source but not always!

spier replied:
While I am not sold on the _channels-are-more-obvious_ argument either, I would just leave it as is for now and see what happens when we get feedback for the pattern from a wider audience. 

Resolving this thread.

gruetter commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,40 @@
+
+## Resolution
+* Corporate requirement that internally developed components should be evaluated before we go outside to search for open source component.
+If the open source component is now more mature, replace the internally developed with the open source. Create some extrinsic rewards to encourage them (initially).
```
I think this needs to be qualified. It the OSS component has strong copyleft, it might make sense to re-engineer its relevant parts in an IS project.

gruetter replied:
On 2nd thought, I think we should remove this bullet point, as it doesn't really solve the underlying problem.

NewMexicoKid commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,40 @@
+* Create a tool with a central view of internally available software, but be aware that people are more inclined to google externally than look internally.
+
+## Resulting Context
+Developers do select the best component regardless of whether it is open source or internally developed. Internally developed software components are then more widely reused.
```
Suggest: Developers select the best component regardless of whether it is open source or internally developed, but they do so with full knowledge of the inner source component, its capabilities, and reasons why it is recommended. Internally developed software components are then more widely and consistently used within the company.

spier replied:
This has been worked into the pattern already. Resolving.

gruetter commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,40 @@
+Make the process simple and aligned with well known open source methods.
+* Provide an internal instance of GitHub Enterprise or an well publicized external GitHub organization repo to allow employees to easily find internally developed solutions.
+Assign maintainers to make sure proper open source processes are being followed within the leading repos.
+Provide “value add” services linked to the formal repo solution such as automated CI/CD services, IaaS/PaaS, NPM organization/server linkages, ChatOps, etc.
```
> "formal repo solution"

Am I right assuming that you mean the mandate by governance to evaluate InnerSource components, first, @NewMexicoKid ? If so, I propose to change to

"Make InnerSource components more attractive by providing "value add" services, such as ...

spier replied:
This sentence has already been changed to:
> Increase the attractiveness of InnerSource components by providing “value add” services for them, such as automated CI/CD services, IaaS/PaaS, NPM organization/server linkages, ChatOps, etc.

Resolving.

gruetter commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,40 @@
+* Provide an internal instance of GitHub Enterprise or an well publicized external GitHub organization repo to allow employees to easily find internally developed solutions.
+Assign maintainers to make sure proper open source processes are being followed within the leading repos.
+Provide “value add” services linked to the formal repo solution such as automated CI/CD services, IaaS/PaaS, NPM organization/server linkages, ChatOps, etc.
+* Create a tool with a central view of internally available software, but be aware that people are more inclined to google externally than look internally.
```
I feel this is a bit of an overlap to the 2nd bullet point here (Make it easier to find ...). I propose to merge these two bullet points into one and link to the "InnerSource Portal" pattern.

spier replied:
Agreed. I have removed the point
> * Provide an internal instance of GitHub Enterprise or a well publicized external GitHub organization repo to allow employees to easily find internally developed solutions.

gruetter committed:
followed up on PR suggestions #46

gruetter committed:
Merge pull request #216 from InnerSourceCommons/oss-trumps-is

Followed up on review suggestions and reworked slightly

gruetter reviewed (approved):
Pattern looks good now and ready to merge, IMHO. I have not yet heard of a company which successfully adopted the core of the solution (mandate considering IS SW before OSS). The pattern conforms to the common structure, though. Therefore, I think the proper maturity level is _2-structured_, @spier , @lenucksi . I'll update the PR tag accordingly.

lenucksi commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,43 @@
+Developers select the best component regardless of whether it is open source or internally developed, but they do so with full knowledge of the inner source component, its capabilities, and reasons why it is recommended. Internally developed software components are then more widely and consistently used within the company.
+
+## Author
+As told to Padma Sudarsan, 2016-09-30
```
We've had quite a few discussions and edits around this. Are there suggestions on who should be included as co-authors to this?

spier replied:
I added all of you as. you seemed to have brought this pattern to where it is today.

new author section
```
* As told to Padma Sudarsan, 2016-09-30
* Tim Yao
* Georg Gruetter (Robert Bosch GmbH)
* Russ Rutledge
* Johannes Tigges
```

lenucksi commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,43 @@
+
+## Patlet
+
+People find the InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with the open source component.
```
I've slightly changed the phrasing here to make it easier to understand. This reflects my understanding of the rest of the pattern. Please correct if I got the point of the pattern wrong.

```suggestion
People find an InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with an equivalent open source component.
```

spier replied:
That's also how I understand this pattern. Adding your suggestion to the PR.

lenucksi commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,43 @@
+People find the InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with the open source component.
+
+## Statement of Problem
+* People find the InnerSource project but, even if the InnerSource component meets their needs better than the respective open source component, they choose the open source component over the InnerSource component.
```
This is pretty much equivalent to the **Patlet** can we deduplicate or remove parts here?

spier replied:
I removed the duplication by writing a new patlet that contains both problem and solution.
Resolving.

lenucksi commented on open-source-trumps-innersource.md:
```diff
@@ -0,0 +1,43 @@
+## Resolution
+* Corporate governance mandates that internally developed components are to be evaluated before open source component are selected.
+* Make it easier to find the internal component (See [InnerSource Portal](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/master/patterns/2-structured/innersource-portal.md)). 
+* Provide an internal instance of GitHub Enterprise or a well publicized external GitHub organization repo to allow employees to easily find internally developed solutions.
```
I think this could be phrased clearer, I do not get the full point here. Do you refer to list of suggested OSS / InnerSource solutions? Possibly along the lines of an internal tech radar? Should the external (publicly visible?) GitHub org also list InnerSource solutions?

spier replied:
Based on a different comment I removed the "Provide an internal instance of GitHub Enterprise ..." bullet as it seemed redundant with the "Make it easier to find ..." one.
Resolving.

lenucksi reviewed (commented):
Thanks a lot for the review and work towards resolving open discussions @gruetter, @NewMexicoKid and @rrrutledge!

> Pattern looks good now and ready to merge, IMHO. I have not yet heard of a company which successfully adopted the core of the solution (mandate considering IS SW before OSS). The pattern conforms to the common structure, though. Therefore, I think the proper maturity level is 2-structured, @spier , @lenucksi . I'll update the PR tag accordingly.

The contribution handbook has this as a verification requirement for `2-structured`: 

> Is validated by at least one known instance
>  - Alternatively: key elements of the pattern have been validated in separate contexts and, in consequence, it is justified to believe the full solution will function

What are the key elements that have been validated in separate contexts here if there is no successful instance of deployment?

I've also added a few review comments and commit suggestion after reading the current result. This is a very interesting pattern. 😄 

And as a last element: Independent of this being merged as `2-structured` or `1-initial` we need a follow-up pull request changing moving the file to the according directory structure.

spier committed:
Update open-source-trumps-innersource.md

Co-authored-by: Johannes Tigges <lenucksi@users.noreply.github.com>

spier committed:
Merge remote-tracking branch 'origin/master' into patch/open-source-trumps-innersource

spier committed:
Moving pattern to 1-initial

spier committed:
fix linter issues. adapting pattern to template.

spier committed:
Splitting up Resulting Context into bullets to improve readability. Also call out InnerSource more explicitly.

spier committed:
make link to other pattern relative rather than absolute

spier committed:
Working in feedback from various threads on the PR.

spier committed:
Adding a new patlet that contains both problem and solution.

spier committed:
Moving pattern to the Initial section in the README. also linking to file instead of PR, and using new Patlet.

spier committed:
renamign Problem section. Adding empty Known Instances section to clearly indicate that there are none yet. Adding myself as co-author :)

spier committed:
spelling out 'stuff'

spier committed:
Merge branch 'master' into patch/open-source-trumps-innersource

spier committed:
Found the original of the sketch used in this pattern. yay

spier committed:
removing now-uncessesary sub-sections in the README

spier committed:
adding IEEE paper to the references section
